cocotb/share/lib/simulator/simulatormodule.c

```diff
@@ -956,11 +959,152 @@ static void add_module_constants(PyObject* simulator)
     rc |= PyModule_AddIntConstant(simulator, "OBJECTS",       GPI_OBJECTS);
     rc |= PyModule_AddIntConstant(simulator, "DRIVERS",       GPI_DRIVERS);
     rc |= PyModule_AddIntConstant(simulator, "LOADS",         GPI_LOADS);
+    rc |= PyModule_AddIntConstant(simulator, "BFM_UI_PARAM",  GpiBfmParamType_Ui);
+    rc |= PyModule_AddIntConstant(simulator, "BFM_SI_PARAM",  GpiBfmParamType_Si);
+    rc |= PyModule_AddIntConstant(simulator, "BFM_STR_PARAM", GpiBfmParamType_Str);
 
     if (rc != 0)
         fprintf(stderr, "Failed to add module constants!\n");
 }
 
+/**
+ * bfm_get_count()
+ *
+ * Returns the number of BFMs registered with CoCoTB
+ */
+static PyObject *bfm_get_count(PyObject *self, PyObject *args) {
+	return PyLong_FromUnsignedLong(cocotb_bfm_num_registered());
+}
+
+/**
+ * bfm_get_info()
+ *
+ * Returns information about a specific BFM
+ */
+static PyObject *bfm_get_info(PyObject *self, PyObject *args) {
+	int id = 0; // TODO:
+
+    if (!PyArg_ParseTuple(args, "i", &id)) {
+        return NULL;
+    }
+
+	return PyTuple_Pack(3,
+			Py_BuildValue("s", cocotb_bfm_typename(id)),
+			Py_BuildValue("s", cocotb_bfm_instname(id)),
+			Py_BuildValue("s", cocotb_bfm_clsname(id))
+			);
```

Comment:
Safer as:
```
return Py_BuildValue("sss",
    cocotb_bfm_typename(id),
    cocotb_bfm_instname(id)),
    cocotb_bfm_clsname(id)
);
```
which handles errors correctly.

Reply:
Appreciate the tip! Done.